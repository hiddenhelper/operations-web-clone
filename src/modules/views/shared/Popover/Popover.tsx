import React, { memo } from 'react';

import Paper from '@material-ui/core/Paper';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Divider from '@material-ui/core/Divider';
import Popper, { PopperPlacementType } from '@material-ui/core/Popper';

import { useStyles } from '../MenuPopover/styles';

export interface IMenuItem {
  title: string;
  callback?: (e: React.MouseEvent<HTMLLIElement>) => void;
}
export interface IPopoverProps {
  menuOptionList: IMenuItem[];
  popoverText?: string;
  placement: PopperPlacementType;
  divider?: boolean;
  isOpen?: boolean;
  buttonRef?: any;
  styleClass?: string;
  onClose?: () => void;
}

const Popover = ({ menuOptionList, popoverText = '', placement, divider = false, styleClass = '', isOpen, onClose, buttonRef }: IPopoverProps) => {
  const classes = useStyles();

  return (
    <div className={isOpen ? classes.iconActive : classes.iconInactive} data-testid="popover-wrapper">
      <Popper
        style={{ zIndex: 999999 }}
        className={styleClass}
        open={isOpen}
        anchorEl={buttonRef?.current}
        role={undefined}
        placement={placement}
        transition={false}
        disablePortal={false}
      >
        <Paper>
          <ClickAwayListener onClickAway={onClose}>
            <MenuList autoFocusItem={isOpen} id="menu-list-grow">
              {menuOptionList.map((node, index) => (
                <div key={index}>
                  {divider && index === menuOptionList.length - 1 && <Divider className={classes.itemDivider} />}
                  <MenuItem
                    data-testid="popover-menu-button"
                    className={classes.buttonWrapper}
                    onClick={e => {
                      node.callback(e);
                      onClose();
                    }}
                    disableRipple={true}
                  >
                    {node.title}
                  </MenuItem>
                </div>
              ))}
            </MenuList>
          </ClickAwayListener>
        </Paper>
      </Popper>
      {popoverText}
    </div>
  );
};

export default memo(Popover);
