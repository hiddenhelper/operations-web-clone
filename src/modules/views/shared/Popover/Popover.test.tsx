import React from 'react';
import { render, RenderResult } from '@testing-library/react';

import Popover, { IPopoverProps } from './Popover';

describe('Popover', () => {
  let wrapper: RenderResult;
  let props: IPopoverProps;

  beforeEach(() => {
    props = {
      menuOptionList: [
        { title: 'Item 1', callback: jest.fn() },
        { title: 'Item 2', callback: jest.fn() },
      ],
      isOpen: false,
      onClose: jest.fn(),
      placement: 'bottom-end',
      divider: false,
    };

    wrapper = render(<Popover {...props} />);
  });

  it('should render', () => {
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with secondary values', () => {
    props.divider = true;

    wrapper = render(<Popover {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
