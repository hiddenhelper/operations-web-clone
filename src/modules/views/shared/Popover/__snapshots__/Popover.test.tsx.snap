// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render 1`] = `
<body>
  <div>
    <div
      class="makeStyles-iconInactive-1"
      data-testid="popover-wrapper"
    >
      
    </div>
  </div>
</body>
`;

exports[`Popover should render with secondary values 1`] = `
<body>
  <div>
    <div
      class="makeStyles-iconInactive-5"
      data-testid="popover-wrapper"
    >
      
    </div>
  </div>
  <div>
    <div
      class="makeStyles-iconInactive-5"
      data-testid="popover-wrapper"
    >
      
    </div>
  </div>
</body>
`;
